File: 371. Sum of Two Integers.cpp
#include <iostream>
#include <bitset>

using namespace std;

class Solution {
public:
    int getSum(int a, int b) {
        int tmp;
        while (a) {
            // cout << bitset<sizeof(int)*8>(a) << ' ' << bitset<sizeof(int)*8>(b) << ' '
            //      << bitset<sizeof(int)*8>(a^b) << ' ' << bitset<sizeof(int)*8>((unsigned int)(a & b) << 1) << endl;
            // cout << a << ' ' << b << ' ' << (a^b) << ' ' << ((a & b) << 1) << endl;
            tmp = a ^ b;
            // a = ((a & b) & 0x7fffffff) << 1;
            a = (unsigned int)(a & b) << 1;
            b = tmp;
        }
        return b;
    }
};

int main() {
    Solution sol;
    cout << sol.getSum(-23, 45) << endl;
    return 0;
}

File: Python3/371. Sum of Two Integers.py
class Solution:
    def plus(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        if a == 0:
            return b
        if b == 0:
            return a
        while b != 0:
            c = a & b
            a = a ^ b
            b = c << 1
        return a

    def minus(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        if a == 0:
            return -b
        if b == 0:
            return a
        while b != 0:
            c = ~a & b
            a = a ^ b
            b = c << 1
        return a

    def getSum(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        if a == 0:
            return b
        if b == 0:
            return a
        if a > 0 and b > 0:
            return self.plus(a, b)
        if a < 0 and b < 0:
            return -self.plus(-a, -b)
        if a > 0:
            if a > -b:
                return self.minus(a, -b)
            else:
                return -self.minus(-b, a)
        if -a < b:
            return self.minus(b, -a)
        else:
            return -self.minus(-a, b)


sol = Solution()
print(sol.getSum(-23, 56))
print(sol.getSum(-1, -1))
print(sol.getSum(134, -1))
print(sol.getSum(12, 341))
print(sol.getSum(12, -341))
print(sol.getSum(-341, 12))
print(sol.getSum(341, -12))
print(sol.getSum(-341, -12))
print(sol.getSum(2, -2))
print(sol.getSum(-2, 2))
